components/transfer/index.tsx

```diff
@@ -148,19 +148,52 @@ const Transfer = <RecordType extends TransferItem = TransferItem>(
   const [wrapSSR, hashId] = useStyle(prefixCls);
 
   const [sourceSelectedKeys, setSourceSelectedKeys] = useState<string[]>(() =>
-    selectedKeys.filter((key) => !targetKeys.includes(key)),
+    (selectedKeys ?? []).filter((key) => !targetKeys.includes(key)),
   );
 
   const [targetSelectedKeys, setTargetSelectedKeys] = useState<string[]>(() =>
-    selectedKeys.filter((key) => targetKeys.includes(key)),
+    (selectedKeys ?? []).filter((key) => targetKeys.includes(key)),
+  );
+
+  const propSourceKeys = useMemo<string[]>(
+    () =>
+      dataSource.filter((item) => !props.targetKeys?.includes(item.key!)).map((item) => item.key!),
+    [dataSource, props.targetKeys],
   );
 
-  useEffect(() => {
-    if (props.selectedKeys) {
-      setSourceSelectedKeys(() => selectedKeys.filter((key) => !targetKeys.includes(key)));
-      setTargetSelectedKeys(() => selectedKeys.filter((key) => targetKeys.includes(key)));
+  const getMergedKeys = (
+    currentTargetKeys: string[],
+    typeSelectedKeys: string[],
+    propTypeKeys: string[] | undefined,
+    propSelectedKeys: string[] | undefined,
+    isSource: boolean,
+  ) => {
+    const keysWithSelectedOrNoTarget = (selectedKeys ?? []).filter((key) =>
+      isSource ? !currentTargetKeys.includes(key) : currentTargetKeys.includes(key),
+    );
+    if (propSelectedKeys) {
+      return keysWithSelectedOrNoTarget;
     }
-  }, [props.selectedKeys, props.targetKeys]);
+    if (propTypeKeys) {
+      return propTypeKeys.filter((key) => typeSelectedKeys.includes(key));
+    }
+    return keysWithSelectedOrNoTarget;
+  };
+
+  const mergedSourceSelectedKeys = useMemo(
+    () => getMergedKeys(targetKeys, sourceSelectedKeys, propSourceKeys, selectedKeys, true),
+    [
+      selectedKeys?.join(''),
```

Comment:
```suggestion
      selectedKeys?.join(','),
```

加个分隔符，否则 `[12, 3]` 和 `[1, 23]` 分不开。

Reply:
done